src/injected/all-frame-runner.ts

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { HTMLElementUtils } from 'common/html-element-utils';
+import { WindowUtils } from 'common/window-utils';
+import { FrameMessenger } from 'injected/frameCommunicators/frame-messenger';
+import {
+    CommandMessage,
+    CommandMessageResponse,
+} from 'injected/frameCommunicators/respondable-command-message-communicator';
+
+export interface AllFrameRunnerTarget<T> {
+    start: () => void;
+    teardown: () => void;
+    transformChildResultForParent: (fromChild: T, messageSourceFrame: HTMLIFrameElement) => T;
+    setResultCallback: (reportResults: (payload: T) => Promise<null>) => void;
+}
+
+/*
+This class ('runner' below) runs methods in AllFrameRunnerTargets ('target' below) in all child
+frames. Runner manages communication between child frames and the top-level window so that the
+target can produce results without explicit frame-communication code. 
+
+It follows these semantics:
+- runner.initialize() should be called & runner.topWindowCallback should be set 
+before calling runner.start() / runner.stop()
+- runner.start() will run target.start() in the current frame, and then in any child frames
+- runner.stop() will run target.stop() in the current frame, and then in any child frames
+- runner provides target with a reportResults callback. When target.reportResults(payload) is 
+called, runner propogates payload to parent frames. In each parent frame, runner replaces the 
+payload with target.transformChildResultForParent(payload). When the result reaches the top window, 
+runner calls topWindowCallback(payload)
+*/
+export class AllFrameRunner<T> {
+    public static readonly startCommand = 'insights.startAllFrameRunner';
+    public static readonly stopCommand = 'insights.stopAllFrameRunner';
+    public static readonly onResultFromChildFrame = 'insights.onResultFromChildInFrameRunner';
+    public topWindowCallback: (result: T) => void;
+
+    constructor(
+        private readonly frameMessenger: FrameMessenger,
+        private readonly htmlElementUtils: HTMLElementUtils,
+        private readonly windowUtils: WindowUtils,
+        private readonly listener: AllFrameRunnerTarget<T>,
+    ) {}
+
+    public initialize() {
+        this.frameMessenger.addMessageListener(AllFrameRunner.startCommand, this.start);
+        this.frameMessenger.addMessageListener(AllFrameRunner.stopCommand, this.stop);
+        this.frameMessenger.addMessageListener(
+            AllFrameRunner.onResultFromChildFrame,
+            this.onResultFromChildFrame,
+        );
+
+        this.listener.setResultCallback(async payload => {
+            await this.reportResults(payload);
+            return null;
+        });
+    }
+
+    public start = async (): Promise<CommandMessageResponse | null> => {
```

Comment:
start & stop are explicitly listed here and in AllFrameRunnerTarget - this feels OK to me, but I briefly considered an 'operations' type that maps from string IDs ("start" | "stop", for instance) to functions. This would remove the similar code between start/stop and enable arbitrary operations defined by AllFrameRunnerTarget. It felt premature so I left it out.